Trustroots/trustroots
# Shadowban

- Adds extra protection by filtering out suspended users.
- Adds a new "shadowban" role which is like suspended, but works slightly differently.

TODO:
- tests covering message thread/inbox changes
- tests covering notification changes
- tests covering profile description changes
- split into small easy-to-review PRs:
   - https://github.com/Trustroots/trustroots/pull/1180
   - https://github.com/Trustroots/trustroots/pull/1179
   - https://github.com/Trustroots/trustroots/pull/1181
   - https://github.com/Trustroots/trustroots/pull/1182
   - https://github.com/Trustroots/trustroots/pull/1183

### To test

- Create two users and log in different browser windows (e.g. use incognito). Confirm emails to both users (e.g. via http://localhost:1080 or switching `public` value to `true` in their profiles).

- Add `shadowban` to another user's `roles` array in the `users` DB

- Confirm that shadowban user can look at their own profile and other people's profiles

- Confirm that other users don't see shadow users profile

- Shadow users can look at other people's offers on the map. Likewise, regular users will see shadow users offers on the map. Hiding these can be done in follow-ups.

- Confirm that shadowban user _can_ send messages to others and can see those message threads

- Confirm that regular users cannot view messages from the shadowban user. Unfortunately, they still see the thread in their inbox and one "unread message". This is identical to how it works with "suspended" users, we'll have to work on that in follow-ups anyway.

    <img width="1210" alt="Screenshot 2020-01-18 at 17 05 13" src="https://user-images.githubusercontent.com/87168/72666167-59e49480-3a18-11ea-9048-c455ca772349.png">
    <img width="1097" alt="Screenshot 2020-01-18 at 17 26 33" src="https://user-images.githubusercontent.com/87168/72666168-5cdf8500-3a18-11ea-8450-11a66939e2df.png">

- Confirm that shadowban user cannot remove themself, while a regular user can

   ![Screenshot 2020-01-18 at 17 11 42](https://user-images.githubusercontent.com/87168/72666172-64069300-3a18-11ea-983b-485e6181594d.png)

- Add `suspended` role and confirm that you cannot login with that user

- Add `admin` role and confirm you can login and see  admin dash at `http://localhost:3000/admin`

- Write some phone numbers, emails, and URLs in the profile description of a regular user. Regular uers see those just fine, but shadowbanned user won't:
 
     <img src="https://user-images.githubusercontent.com/87168/72666800-4805f000-3a1e-11ea-9f53-a6916c301def.png" alt="" width="350">
    <img width="350" alt="Screenshot 2020-01-18 at 18 10 26" src="https://user-images.githubusercontent.com/87168/72666793-44726900-3a1e-11ea-94b9-b542ce126dd4.png">



- I also modified notifications behavior for shadowban + suspended users but I have to write testing instructions for that. :-)
